File: ternary_search.js
function ternarySearchRecursive(arr, key, low = 0, high = arr.length - 1) {
  if (high >= low) {
    const mid1 = Math.floor(low + (high - low) / 3)
    const mid2 = Math.floor(high - (high - low) / 3)

    if (arr[mid1] === key) {
      return mid1
    }
    if (arr[mid2] === key) {
      return mid2
    }

    if (key < arr[mid1]) {
      return ternarySearchRecursive(arr, key, low, mid1 - 1)
    } else if (key > arr[mid2]) {
      return ternarySearchRecursive(arr, key, mid2 + 1, high)
    } else {
      return ternarySearchRecursive(arr, key, mid1 + 1, mid2 - 1)
    }
  } else {
    return -1
  }
}

function ternarySearchIterative(arr, key, low = 0, high = arr.length - 1) {
  while (high >= low) {
    const mid1 = Math.floor(low + (high - low) / 3)
    const mid2 = Math.floor(high - (high - low) / 3)

    if (arr[mid1] === key) {
      return mid1
    }
    if (arr[mid2] === key) {
      return mid2
    }

    if (key < arr[mid1]) {
      high = mid1 - 1
    } else if (key > arr[mid2]) {
      low = mid2 + 1
    } else {
      low = mid1 + 1
      high = mid2 - 1
    }
  }
  return -1
}

export { ternarySearchRecursive, ternarySearchIterative }

File: ternary_search.rb
def ternary_search(l, r, key, arr)
  if r >= l
      mid1 = l + (r - l) / 3
      mid2 = r - (r - l) / 3
      if arr[mid1] == key
          mid1
      elsif arr[mid2] == key
          mid2
      elsif key < arr[mid1]
          ternary_search(l, mid1 - 1, key, arr)
      elsif key > arr[mid2]
          ternary_search(mid2 + 1, r, key, arr)
      else
          ternary_search(mid1 + 1, mid2 - 1, key, arr)
      end
  end
end

puts "Enter a space-separated list:"
arr = gets.chomp.split(' ').map(&:to_i)
puts "Enter a value to be searched:"
key = gets.chomp.to_i
puts if ternary_search(0, arr.length - 1, key, arr) != nil
  "Found at index #{ternary_search(0, arr.length - 1, key, arr)}"
else
  "Not found"
end
